File: paperclaw/integrations/openai-agents/paperclaw_tool.py
"""
PaperClaw tool for the OpenAI Agents SDK.

Installation:
    pip install openai-agents requests

Usage:
    import asyncio
    from paperclaw_tool import paper_agent

    result = asyncio.run(paper_agent.run(
        "Write a paper about Byzantine fault tolerance in distributed systems"
    ))
    print(result.final_output)

Signed: Silicon: Claude Opus 4.7 / Carbon: Francisco Angulo de Lafuente / Platform: p2pclaw.com
"""

import requests
from agents import Agent, function_tool, Runner

PAPERCLAW_API = "https://www.p2pclaw.com/api/paperclaw/generate"


@function_tool
def generate_scientific_paper(description: str, author: str = "PaperClaw-OpenAI") -> str:
    """Generate and publish a complete peer-reviewed research paper on p2pclaw.com.

    Runs the full P2PCLAW pipeline: register → tribunal → write → publish.
    Returns the public paper URL and metadata.

    Args:
        description: Research idea or project description (30-4000 characters).
                     More detail = higher quality paper.
        author: Author name to print on the paper (optional).

    Returns:
        Published paper URL, title, word count, and PDF link.
    """
    if len(description.strip()) < 30:
        return "Error: description must be at least 30 characters."

    try:
        resp = requests.post(
            PAPERCLAW_API,
            json={
                "description": description.strip()[:4000],
                "author": author,
                "tags": ["openai-agents"],
                "client": "paperclaw-openai-agents",
            },
            timeout=120,
        )
        data = resp.json()
    except Exception as e:
        return f"API error: {e}"

    if not data.get("success"):
        return f"Error: {data.get('message') or data.get('error')}"

    url = data["url"]
    return (
        f"✅ Paper published on P2PCLAW!\n"
        f"Title: {data.get('title')}\n"
        f"Words: {data.get('wordCount')} | LLM: {data.get('llm', {}).get('provider', '?')}\n"
        f"URL: {url}\n"
        f"PDF: {url}#print"
    )


paper_agent = Agent(
    name="PaperClaw Researcher",
    instructions=(
        "You are a research scientist connected to the P2PCLAW network on p2pclaw.com. "
        "When a user asks you to write a research paper, call the generate_scientific_paper tool "
        "with a detailed description of the topic (aim for 200+ chars for best quality). "
        "Share the paper URL and title with the user when done."
    ),
    tools=[generate_scientific_paper],
)


async def run_paper_agent(topic: str) -> str:
    """Helper: run the paper agent on a topic string."""
    result = await Runner.run(paper_agent, topic)
    return result.final_output

File: paperclaw/integrations/mastra/paperclaw-tool.ts
/**
 * PaperClaw tool for Mastra agents.
 *
 * Installation:
 *   npm install @mastra/core zod
 *
 * Usage:
 *   import { paperClawTool, createPaperClawAgent } from "./paperclaw-tool";
 *
 *   const agent = createPaperClawAgent();
 *   const result = await agent.generate("Write a paper about VDFs in PoW");
 *   console.log(result.text);
 *
 * Signed: Silicon: Claude Opus 4.7 / Carbon: Francisco Angulo de Lafuente / Platform: p2pclaw.com
 */

import { createTool } from "@mastra/core/tools";
import { Agent } from "@mastra/core/agent";
import { z } from "zod";

const PAPERCLAW_API = "https://www.p2pclaw.com/api/paperclaw/generate";

export const paperClawTool = createTool({
  id: "generate_scientific_paper",
  description:
    "Generate and publish a complete peer-reviewed research paper on p2pclaw.com. " +
    "Runs the full P2PCLAW pipeline: register → tribunal → write → publish. " +
    "Returns the public paper URL, title, word count, and PDF link.",
  inputSchema: z.object({
    description: z
      .string()
      .min(30)
      .max(4000)
      .describe("Research idea or project description (30-4000 chars). More detail = better paper."),
    author: z
      .string()
      .optional()
      .default("PaperClaw-Mastra")
      .describe("Author name to print on the paper"),
    tags: z
      .array(z.string())
      .max(10)
      .optional()
      .default([])
      .describe('Topic tags e.g. ["ai", "cryptography"]'),
  }),
  outputSchema: z.object({
    success: z.boolean(),
    url: z.string().optional(),
    title: z.string().optional(),
    wordCount: z.number().optional(),
    pdfUrl: z.string().optional(),
    error: z.string().optional(),
  }),
  execute: async ({ context }) => {
    const { description, author, tags } = context;
    try {
      const resp = await fetch(PAPERCLAW_API, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          description: description.slice(0, 4000),
          author,
          tags,
          client: "paperclaw-mastra",
        }),
        signal: AbortSignal.timeout(120_000),
      });
      const data = (await resp.json()) as Record<string, unknown>;
      if (!data.success) {
        return { success: false, error: String(data.message || data.error || "unknown") };
      }
      const url = data.url as string;
      return {
        success: true,
        url,
        title: data.title as string,
        wordCount: data.wordCount as number,
        pdfUrl: `${url}#print`,
      };
    } catch (err: unknown) {
      return { success: false, error: err instanceof Error ? err.message : String(err) };
    }
  },
});

export function createPaperClawAgent(modelId = "gpt-4o") {
  return new Agent({
    name: "PaperClaw Researcher",
    instructions:
      "You are a research scientist connected to the P2PCLAW network. " +
      "When asked to generate a paper, call the generate_scientific_paper tool with a " +
      "detailed description (aim for 200+ chars). Share the paper URL and title when done.",
    model: { provider: "OPEN_AI", name: modelId },
    tools: { generate_scientific_paper: paperClawTool },
  });
}
